swarm/storage/mru/topic.go

```diff
@@ -0,0 +1,56 @@
+package mru
+
+import (
+	"bytes"
+	"encoding/json"
+
+	"github.com/ethereum/go-ethereum/common/bitutil"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/swarm/storage"
+)
+
+const topicLength = storage.KeyLength
+
+type Topic struct {
+	content [topicLength]byte
+}
+
+func NewTopic(name string, relatedContent storage.Address) Topic {
+	var topic Topic
+	copy(topic.content[:], relatedContent)
+	nameLength := len(name)
+	if nameLength > topicLength {
+		nameLength = topicLength
+	}
+	bitutil.XORBytes(topic.content[:], topic.content[:], []byte(name)[:nameLength])
+	return topic
+}
+
+func (t *Topic) Hex() string {
+	return hexutil.Encode(t.content[:])
+}
+
+func (t *Topic) FromHex(hex string) error {
+	return decodeHexArray(t.content[:], hex, "Topic")
+}
+
+func (t *Topic) Name(relatedContent storage.Address) string {
+	nameBytes := t
+	bitutil.XORBytes(nameBytes.content[:], t.content[:], relatedContent)
+	z := bytes.IndexByte(nameBytes.content[:], 0)
+	if z < 0 {
+		z = topicLength
+	}
```

Comment:
So you need `bitutil` for some fuzzy filter alchemy, maybe? Could you explain, please?

Reply:
Alchemy! I love that word applied to this concept.
The trick here is that your resource topic can come from either of the following:
1. A string name, e.g., `world news`
2. A *relatedContent*. This is a 32-byte `storage.Address`. An arbitrary 32-byte array, or a nonce. It is called `relatedContent` because it could be the content address of something this MRU is talking about.
3. A combination of a string name and a *relatedContent*. In this case, we XOR the two together to create a new unique topic. We are using XOR here as a very cheap hash.

For (2) and (3) if `relatedContent` is for example the hash of a picture, then the MRU could mean it is mutable information about that particular picture.

For (3), you could have various topics for the same picture. For example, if `relatedContent` is the hash of the picture, and `name == "comments"` then the topic is `XOR(relatedContent, "comments")`, thus creating a new unique topic where people's comments about that particular picture are collected.

This is an implementation of @zelig's idea of a "disqus"-like MRU. It turned out it came in pretty nicely.

If `relatedContent` is `nil`, then the topic comes just from the name (`XOR(0, name) = name`), and the name can be retrieved. The name is limited to 32 characters, though, but that is plenty.

If `name == ""` and `relatedContent==nil` then the topic is zero `0x0000...`, which is not bad either. This could be a way of having a default topic where you can look up information about a user, for example.

Reply:
Very nice!

Reply:
well you are guaranteed to collide if the human readable parts share a 32 byte prefix...

Reply:
There would only be one human-readable part. In case there is a collision, maybe that is what you are actually looking for, your meeting point.